ta4j-core/src/main/java/org/ta4j/core/indicators/RandomWalkIndexHighIndicator.java

```diff
@@ -54,14 +52,28 @@ public RandomWalkIndexHighIndicator(TimeSeries series, int timeFrame) {
         this.timeFrame = timeFrame;
         maxPrice = new MaxPriceIndicator(series);
         minPrice = new MinPriceIndicator(series);
-        averageTrueRange = new AverageTrueRangeIndicator(series, timeFrame);
-        sqrtTimeFrame = Decimal.valueOf(timeFrame).sqrt();
+
     }
 
     @Override
```

Comment:
**Please replace your calculate()-method to something like this:**
```

@Override
	protected Decimal calculate(int index) {
		
		Decimal highestRWI = Decimal.NaN;
		int startIndex = Math.max(0, index - timeFrame + 1);
		
		for (int i = startIndex; i <= index - 1; i++) {
			Decimal currentRWI = calculateRwiHigh(index, i);
			if (currentRWI.isGreaterThan(highestRWI) || highestRWI.isNaN()) {
				highestRWI = currentRWI;
			}
		}
		
		return highestRWI;
	}
	
	private Decimal calculateRwiHigh(int index, int pos){
		AverageTrueRangeIndicator atr = new AverageTrueRangeIndicator(getTimeSeries(), pos);
		return maxPrice.getValue(index).minus(minPrice.getValue(pos))
				.dividedBy(atr.getValue(index).multipliedBy(Decimal.valueOf(Math.sqrt(pos))));
	}
```

Reply:
`maxPrice.getValue(index).minus(minPrice.getValue(pos)` -> `pos` must **starting with index - one** (today-yesterday) and gets incremented in each iteration by one.

The time frame of the ATR (**starting with two**) must be incremented by one in each iteration. In your version it starts at startIndex and ends with currentIndex ??

Reply:
for example, timeframe=4

it goes back 4 times and starts there (=startIndex), calculates each rwi till actual index-1 (because the actual index needs at least one previous bar/tick) and then returns the highest of all the 4 calculated rwi.

However, take your loop with start=2 and decrement (go back till index-timeframe) instead of increment and eliminate the redundancy to the form above.

Reply:
`new AverageTrueRangeIndicator(getTimeSeries(), pos);` pos has to be the timeFame not index

Reply:
https://rtmath.net/helpFinAnalysis/html/934563a8-9171-42d2-8444-486691234b1d.htm

>where Atr is the average true range over N periods preceding the current period. 

ok, then:

private Decimal calculateRwiHigh(int index, int pos){
	AverageTrueRangeIndicator atr = new AverageTrueRangeIndicator(getTimeSeries(), pos);
        // TODO: which to use?: "atr.getValue(index)" or "atr.getValue(timeFrame)"
	return maxPrice.getValue(index).minus(minPrice.getValue(pos))
			.dividedBy(atr.getValue(index).multipliedBy(Decimal.valueOf(Math.sqrt(pos))));
}

which to use?: ".dividedBy(atr.getValue(index))" or ".dividedBy(atr.getValue(timeFrame))"

Reply:
I think `.dividedBy(atr.getValue(index)` because in `AverageTrueRangeIndicator atr = new AverageTrueRangeIndicator(getTimeSeries(), timeFrame);` we have to take the "**current**" timeFrame, _**form `n=2 ` to `this.timeFrame`.**_ So we need in each iteration the index (always same) and the currentTimeFrame (n that increments in each iteration)

Reply:
http://www.neuroshell.com/manuals/ais2/index.html?rndwalkhigh.htm